### ISSUE TITLE
Error restoring with UnicodeEncodeError

### ISSUE THREAD

**Issue by NovisX** [jay0lee/got-your-back#51]
Hi,

I am restoring my old gmail email to a new gmail acocunt. I have come across an error:

UnicodeEncodeError: 'latin-1' codec can't encode character '\udce1'

![error-image](https://cloud.githubusercontent.com/assets/4776528/9327768/a0632c6e-45a3-11e5-8881-bbc0f223323e.jpg)

How can I overcome this? Is there a command I can use to ignore these characters and skip the error?
Many Thanks for any help.

**Comment by fatkero** [comment]
Same problem here. Please help us out! Loads of thanks!

**Comment by jay0lee** [comment]
I need to see some test messages where this is occurring. If you can narrow the backup down to a handful of messages that have this problem and then email me the backup (maybe share via Google Drive) that would go along way towards allowing me to reproduce the issue and see what can be done.

**Comment by NovisX** [comment]
Hi Jay,
No problem, I will send you the 2 months where I have encountered the error. I did a backup 2 times just to test and the errors both occured after the 300th email approximately, but on different emails. The second restore I attempted, went past where the forst error was, but stopped at another point.
Many Thanks for you help!!

**Comment by NovisX** [comment]
Hi Jay,

I have just shared months of emails with you in google Drive. Let me know if you need anything else.
Many Thanks

**Comment by hoagieCodes** [comment]
Hi Jay,
I have the same unicode issues with a number of mails. I wish GotYourBack would just skip these mails and continue uploading (and afterwards show a list of the skipped messages). Do you have enough test messages to reproduce the error? Would love to get your help here! Thanks in advance!

**Comment by otint** [comment]
Same problem. +1

**Comment by SuperSpectre** [comment]
+1.

However, you can nail down the problematic mails by looking into the @gmail.com-restored.sqlite database, which shows you which mails were processed sucessfully. You can copy the highest number from there, look up the .eml-file in the msg-db.sqlite that matches this number and then search for the folder containing the .eml. For me, it were the bigger mails with attachments in the according folder that caused the errors (so the highest index does *not* tell you the exact problematic mail!). You can then rename those problematic files to exclude them from processing.

**Comment by Plazzmex** [comment]
the same problem in here...
How can i skip the mail that is causing the problem ?
I Can't understand the sqlite file...

**Comment by fourjay** [comment]
I can provide a test message if needed.

**Comment by Northgate-GIT** [comment]
@Plazzmex .  You can download this free SQL lite viewer:
http://sqlitebrowser.org/

Then following @SuperSpectre's approach you can narrow down the folder that contains the messages that caused error.

In my specific case, by looking at the messages table in the @gmail.com-restored.sqlite database I was able to identify that the last successful message processed was message #1510.  
I then opened the msg-db.sqlite database and went to the messages table and looked up the message_filename path for message #1510 (e.g. C:\gyb\GYB-GMail-Backup-myemail@gmail.com\NGEM\2010\10\19\12bc3b6326dbf480.eml ) and found that there were two *.eml files that were bigger than 5mb in the folder. I changed the file name of both of these files (for example I just added the word "skip" to end of each file name), and re-ran the restore commands.   GYB gave an alert saying it could not find the two files and said it would skip them, and then things proceeded along fine (for a while at least until I encountered a new error message...).

I suspect that most people using GYB  are close to exceeding their free disk quota and so have close to 10,000  emails with everything under the earth in terms of headers, file sizes, attachments, encoding, etc... so will be hard to get through full stack of messages without getting other errors....

**Comment by samsar777** [comment]
Should you need more cases, I can share few emails with this issue.

**Comment by natanbr** [comment]
+1 here :-(

**Comment by cp-p** [comment]
I think this is somehow related to this bug:
https://github.com/google/google-api-python-client/commit/cd456d578174842e865a0547717bfc6440f9302f

However just changing to the newest google api client did not solve it completely, as another error shows up. Don't have the time to dig into that unfortunately.

I did then change line 1077 in gyb.py from 

`media_body = googleapiclient.http.MediaInMemoryUpload(
full_message,mimetype='message/rfc822')`

to 

`media_body = googleapiclient.http.MediaInMemoryUpload(
full_message,mimetype='message/rfc822',chunksize=1024*1024,resumable=True)`

and this seems to work actually. Have not seen any side effects yet.

**Comment by hyperion4ever** [comment]
+1

**Comment by JonathanCasey** [comment]
I ran into an encoding issue as well.  In my case, the offending .eml file was less than 1MB in size, but the email did have a PDF attachment.  I am not sure if I attempted @cp-p's solution--I was trying a lot of different things rapid fire.

The way I saw it, Python's http lib was defaulting to a latin-1 encoding, which caused the issue when a character in the utf-8 realm was encountered.

To solve this, I inserted the following at [line 720 of googleapiclient/http.py](https://github.com/jay0lee/got-your-back/blob/a6b1c91076c4a0bc7c5e093a3c3281ff5191a094/googleapiclient/http.py#L720):

```python
if self.body:
  self.body = self.body.encode('utf-8', errors='surrogateescape')
```

**Disclaimer**: I am not sure if this is the proper solution in all cases, nor if it is the best place to do the encoding.

This worked for me on this email, as both the email and the PDF seemed flawless after the transfer.  I left this in for all remaining emails--I did not find anything that was transferred incorrectly.  In my case, for the original offending email, I got an error about an unmatched surrogate pair when I only encoded to 'utf-8' with no other arguments, hence I added the surrogateescape argument.  There is some info on that [here](http://python-notes.curiousefficiency.org/en/latest/python3/text_file_processing.html#unicode-error-handlers), which I found through [this](http://stackoverflow.com/questions/21116089/surrogateescape-cannot-escape-certain-characters).

Hope this helps...

**Comment by ratio91** [comment]
Thanks @JonathanCasey - your solution works perfectly fine for me.
So in total, I had to make these 2 changes:

**http.py (line 720):**
`if self.body: self.body = self.body.encode('utf-8', errors='surrogateescape')`

**gyb.py (line 414):**
`json_string = open(oauth2servicefilejson, 'rb').read().decode('utf-8')`

**Comment by subibro** [comment]
I too came across the same issue, inspected the files, not large files like in @JonathanCasey and @ratio91 
Any fix in sight? I am not comfortable making these mods to the code.
Can anyone make the .exe available with these corrections?

**Comment by linedotpaul** [comment]
Narrowed mine down to one particular large file. Also has a pdf attachment so perhaps it's that? The only other dodgy character i noticed was a percentage sign in the pdf file name

**Comment by subibro** [comment]
I took note of the last message restored, found its correspond filename,
went into its directory and did not find there any large file, although it
says "restoring large file" before it comes up with the error messages.
I wonder if I am missing something or if I have a problem with the handling
of special characters, latin1, unicode, etc

**Comment by linedotpaul** [comment]
Subibro what is the email one date after the last one it restored? It's not the one it last restored that is the problem its the one after it in date order. Find that one, delete it. It's almost definitely a large file. If you are not sure which one it is, cut a whole load of emails out of the backup in a date range around the suspicious date. Then perform the backup. Then slowly add those emails back into the backup folder where they should be and try the backup again (it will only do emails not previously restored) until you find the one that is the problem (which will be evident because it will generate the error again.

**Comment by subibro** [comment]
Thanks @linedotpaul. I did not find any large files but a quick test with
Linux 'find' command showed me I have 627 files larger than 5M.
Now the question is, can GYB restore these emails? Any other way around
populating a new account with these emails?

**Comment by apendergrass** [comment]
@cp-p's fix worked for me

**Comment by jay0lee** [comment]
This is fixed in master (soon to be 0.44)
